keras_cv/layers/object_detection/roi_align.py

```diff
@@ -0,0 +1,367 @@
+                boxes[:, :, 0:2],
+                tf.expand_dims(box_height, -1),
+                tf.expand_dims(box_width, -1),
+            ],
+            axis=-1,
+        )
+
+        # Maps levels to [0, max_level-min_level].
+        levels -= min_level
+        level_strides = tf.pow([[2.0]], tf.cast(levels, tf.float32))
+        boundary = tf.cast(
+            tf.concat(
+                [
+                    tf.expand_dims(
+                        [[tf.cast(max_feature_height, tf.float32)]] / level_strides - 1,
+                        axis=-1,
+                    ),
+                    tf.expand_dims(
+                        [[tf.cast(max_feature_width, tf.float32)]] / level_strides - 1,
+                        axis=-1,
+                    ),
+                ],
+                axis=-1,
+            ),
+            boxes.dtype,
+        )
+
+        # Compute grid positions.
+        kernel_y, kernel_x, box_gridy0y1, box_gridx0x1 = _compute_grid_positions(
+            boxes, boundary, output_size, sample_offset
+        )
+
+        x_indices = tf.cast(
+            tf.reshape(box_gridx0x1, [batch_size, num_boxes, output_size * 2]),
+            dtype=tf.int32,
+        )
+        y_indices = tf.cast(
+            tf.reshape(box_gridy0y1, [batch_size, num_boxes, output_size * 2]),
+            dtype=tf.int32,
+        )
+
+        batch_size_offset = tf.tile(
+            tf.reshape(tf.range(batch_size) * batch_dim_size, [batch_size, 1, 1, 1]),
+            [1, num_boxes, output_size * 2, output_size * 2],
+        )
+        # Get level offset for each box. Each box belongs to one level.
+        levels_offset = tf.tile(
+            tf.reshape(
+                tf.gather(level_dim_offsets, levels), [batch_size, num_boxes, 1, 1]
+            ),
+            [1, 1, output_size * 2, output_size * 2],
+        )
+        y_indices_offset = tf.tile(
+            tf.reshape(
+                y_indices * tf.expand_dims(tf.gather(height_dim_sizes, levels), -1),
+                [batch_size, num_boxes, output_size * 2, 1],
+            ),
+            [1, 1, 1, output_size * 2],
+        )
+        x_indices_offset = tf.tile(
+            tf.reshape(x_indices, [batch_size, num_boxes, 1, output_size * 2]),
+            [1, 1, output_size * 2, 1],
+        )
+        indices = tf.reshape(
+            batch_size_offset + levels_offset + y_indices_offset + x_indices_offset,
+            [-1],
+        )
+
+        # TODO(tanzhenyu): replace tf.gather with tf.gather_nd and try to get similar
+        # performance.
+        features_per_box = tf.reshape(
+            tf.gather(features_r2, indices),
+            [batch_size, num_boxes, output_size * 2, output_size * 2, num_filters],
+        )
+
+        # Bilinear interpolation.
+        features_per_box = _feature_bilinear_interpolation(
+            features_per_box, kernel_y, kernel_x
+        )
+        return features_per_box
+
+
+class _ROIAligner(tf.keras.layers.Layer):
```

Comment:
Sometimes I bring in a domain expert to review a component to make sure its correcy (i.e. Ekin for RandAugment, Dan Hendrycks for AugMix).

any domain experts that might be interested in reviewing this?  I unfortunately to not have the background to review this for correctness; only code style.

Reply:
I don't know if @yeqingli could help to check roi aligner.

https://github.com/tensorflow/models/blob/master/official/vision/modeling/layers/roi_aligner.py

Reply:
Ideally we should only use roi pooling in FasterRCNN, so let's keep this as-is until we start supporting MaskRCNN?